### PR TITLE
LIE/HM part 4: Local assembler creation

### PR DESCRIPTION
**Pull request by norihiro-w**
a part of #1535, depends on #1539 

This PR adds creation and initialization of the local assemblers

### CODE REVIEW

**Comment by norihiro-w** [comment]
@endJunction @wenqing could you please review this? I'm sorry for pushing you, because this is my last week...

File: ProcessLib/LIE/HydroMechanics/LocalAssembler/LocalDataInitializer.h [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+            const std::vector<std::size_t> varIDs(_dof_table.getElementVariableIDs(id));
+
+            std::vector<unsigned> dofIndex_to_localIndex;
+            if (mesh_item.getDimension() == GlobalDim
+                && n_global_components == GlobalDim + 1) // +1 for pressure
+            {
+                // normal matrix assembler
+                dofIndex_to_localIndex.resize(n_local_dof);
+                int count = 0;
+                std::generate(dofIndex_to_localIndex.begin(), dofIndex_to_localIndex.end(), [&](){return count++;});
```
Comment (endJunction) [inline ProcessLib/LIE/HydroMechanics/LocalAssembler/LocalDataInitializer.h]: std::iota(dofIndex_to_localIndex.begin(), dofIndex_to_localIndex.end(), 0)

File: ProcessLib/LIE/HydroMechanics/LocalAssembler/LocalDataInitializer.h [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+    void operator()(
+            std::size_t const id,
+            MeshLib::Element const& mesh_item,
+            LADataIntfPtr& data_ptr,
+            ConstructorArgs&&... args) const
+    {
+        auto const type_idx = std::type_index(typeid(mesh_item));
+        auto const it = _builder.find(type_idx);
+
+        if (it != _builder.end())
```
Comment (endJunction) [inline ProcessLib/LIE/HydroMechanics/LocalAssembler/LocalDataInitializer.h]: checking for `it == builder.end()` first, saves one indentation level later.

File: ProcessLib/LIE/HydroMechanics/LocalAssembler/CreateLocalAssemblers.h [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+          class LocalAssemblerFractureImplementation,
+          typename LocalAssemblerInterface, typename... ExtraCtorArgs>
+void createLocalAssemblers(
+    std::vector<MeshLib::Element*> const& mesh_elements,
+    NumLib::LocalToGlobalIndexMap const& dof_table,
+    const unsigned shapefunction_order,
+    std::vector<std::unique_ptr<LocalAssemblerInterface>>& local_assemblers,
+    ExtraCtorArgs&&... extra_ctor_args)
+{
+    DBUG("Create local assemblers.");
```
Comment (norihiro-w) [inline ProcessLib/LIE/HydroMechanics/LocalAssembler/CreateLocalAssemblers.h]: removed the log output

File: ProcessLib/LIE/HydroMechanics/LocalAssembler/LocalDataInitializer.h [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+                bool, (GlobalDim >= ShapeFunctionDisplacement::DIM)>*>(
+                nullptr));
+    }
+
+    /// Mapping of element types to local assembler constructors.
+    std::unordered_map<std::type_index, LADataBuilder> _builder;
+
+    NumLib::LocalToGlobalIndexMap const& _dof_table;
+
+    // local assembler builder implementations.
```
Comment (norihiro-w) [inline ProcessLib/LIE/HydroMechanics/LocalAssembler/LocalDataInitializer.h]: removed the comment

**Comment by endJunction** [comment]
You might want to fix a typo in a commit message with std::iota. iota is mistyped.

**Pull request merged**